File: YFDebugPannel/YFDebugPannel/HCEnvKit.m
#import "HCEnvKit.h"

NSNotificationName const HCEnvKitConfigDidChangeNotification = @"HCEnvKitConfigDidChangeNotification";

static NSString * const HCEnvKitDefaultsKey = @"HCEnvKit.config";
static NSString * const HCEnvKitReleaseBaseURL = @"https://release.example.com";
static NSString * const HCEnvKitUATTemplate = @"https://uat-%ld-%@.example.com"; // TODO: replace with real template
static NSString * const HCEnvKitDevTemplate = @"https://dev-%ld-%@.example.com"; // TODO: replace with real template

@implementation HCEnvConfig

@end

@implementation HCEnvBuildResult

@end

@implementation HCEnvKit

+ (HCEnvConfig *)currentConfig {
    NSDictionary *stored = [[NSUserDefaults standardUserDefaults] dictionaryForKey:HCEnvKitDefaultsKey];
    HCEnvConfig *config = [[HCEnvConfig alloc] init];
    if (stored) {
        config.envType = [stored[@"envType"] integerValue];
        config.clusterIndex = [stored[@"clusterIndex"] integerValue];
        config.isolation = stored[@"isolation"] ?: @"";
        config.version = stored[@"version"] ?: @"v1";
    } else {
        config.envType = HCEnvTypeRelease;
        config.clusterIndex = 1;
        config.isolation = @"";
        config.version = @"v1";
    }
    return config;
}

+ (void)saveConfig:(HCEnvConfig *)config {
    NSDictionary *payload = @{
        @"envType": @(config.envType),
        @"clusterIndex": @(config.clusterIndex),
        @"isolation": config.isolation ?: @"",
        @"version": config.version ?: @"v1"
    };
    [[NSUserDefaults standardUserDefaults] setObject:payload forKey:HCEnvKitDefaultsKey];
    [[NSUserDefaults standardUserDefaults] synchronize];
    [[NSNotificationCenter defaultCenter] postNotificationName:HCEnvKitConfigDidChangeNotification object:nil];
}

+ (HCEnvBuildResult *)buildResult:(HCEnvConfig *)config {
    HCEnvBuildResult *result = [[HCEnvBuildResult alloc] init];
    result.isolation = config.isolation ?: @"";
    if (config.envType == HCEnvTypeRelease) {
        result.displayName = @"线上";
        result.baseURL = HCEnvKitReleaseBaseURL;
        return result;
    }
    NSString *version = config.version.length > 0 ? config.version : @"v1";
    if (config.envType == HCEnvTypeUAT) {
        result.displayName = [NSString stringWithFormat:@"uat-%ld", (long)config.clusterIndex];
        result.baseURL = [NSString stringWithFormat:HCEnvKitUATTemplate, (long)config.clusterIndex, version];
    } else {
        result.displayName = [NSString stringWithFormat:@"dev-%ld", (long)config.clusterIndex];
        result.baseURL = [NSString stringWithFormat:HCEnvKitDevTemplate, (long)config.clusterIndex, version];
    }
    return result;
}

@end

File: YFDebugPannel/YFDebugPannel/HCEnvKit.swift
import Foundation

enum HCEnvType: Int {
    case release
    case uat
    case dev
}

let HCEnvKitConfigDidChangeNotification = Notification.Name("HCEnvKitConfigDidChangeNotification")

final class HCEnvConfig {
    var envType: HCEnvType = .release
    var clusterIndex: Int = 1
    var isolation: String = ""
    var version: String = "v1"
}

final class HCEnvBuildResult {
    var baseURL: String = ""
    var displayName: String = ""
    var isolation: String = ""
}

enum HCEnvKit {
    private static let defaultsKey = "HCEnvKit.config"
    private static let releaseBaseURL = "https://release.example.com"
    private static let uatTemplate = "https://uat-%ld-%@.example.com"
    private static let devTemplate = "https://dev-%ld-%@.example.com"

    static func currentConfig() -> HCEnvConfig {
        let stored = UserDefaults.standard.dictionary(forKey: defaultsKey)
        let config = HCEnvConfig()
        if let stored {
            if let envType = stored["envType"] as? Int {
                config.envType = HCEnvType(rawValue: envType) ?? .release
            }
            if let clusterIndex = stored["clusterIndex"] as? Int {
                config.clusterIndex = clusterIndex
            }
            config.isolation = stored["isolation"] as? String ?? ""
            config.version = stored["version"] as? String ?? "v1"
        } else {
            config.envType = .release
            config.clusterIndex = 1
            config.isolation = ""
            config.version = "v1"
        }
        return config
    }

    static func saveConfig(_ config: HCEnvConfig) {
        let payload: [String: Any] = [
            "envType": config.envType.rawValue,
            "clusterIndex": config.clusterIndex,
            "isolation": config.isolation,
            "version": config.version
        ]
        UserDefaults.standard.set(payload, forKey: defaultsKey)
        UserDefaults.standard.synchronize()
        NotificationCenter.default.post(name: HCEnvKitConfigDidChangeNotification, object: nil)
    }

    static func buildResult(_ config: HCEnvConfig) -> HCEnvBuildResult {
        let result = HCEnvBuildResult()
        result.isolation = config.isolation
        switch config.envType {
        case .release:
            result.displayName = "线上"
            result.baseURL = releaseBaseURL
            return result
        case .uat, .dev:
            break
        }
        let version = config.version.isEmpty ? "v1" : config.version
        if config.envType == .uat {
            result.displayName = "uat-\(config.clusterIndex)"
            result.baseURL = String(format: uatTemplate, config.clusterIndex, version)
        } else {
            result.displayName = "dev-\(config.clusterIndex)"
            result.baseURL = String(format: devTemplate, config.clusterIndex, version)
        }
        return result
    }
}
